src/field_theory/chevalley_warning.lean

```diff
@@ -0,0 +1,195 @@
+  calc (∑ x : {x : σ → K // x ∘ coe = x₀}, ∏ j, (x : σ → K) j ^ d j) =
+                  ∑ a : K, ∏ j : σ, (e a : σ → K) j ^ d j      : (finset.sum_equiv e _).symm
+            ... = ∑ a : K, (∏ j, x₀ j ^ d j) * a ^ d i         : fintype.sum_congr _ _ _
+            ... = (∏ j, x₀ j ^ d j) * ∑ a : K, a ^ d i         : by rw mul_sum
+            ... = 0 : by rw [sum_pow_lt_card_sub_one _ hi, mul_zero],
+  intros a,
+  let e' : {j // j = i} ⊕ {j // j ≠ i} ≃ σ := equiv.sum_compl _,
+  calc (∏ j : σ, (e a : σ → K) j ^ d j) =
+          (e a : σ → K) i ^ d i * (∏ (j : {j // j ≠ i}), (e a : σ → K) j ^ d j) :
+  begin
+    rw [← finset.prod_equiv e', fintype.prod_sum_type],
+    have this : _ = (e a : σ → K) i ^ d i :=
+      @fintype.prod_unique {j // j = i} _ _ _ (λ j, (e a : σ → K) j ^ d j),
+    rw ← this,
+    congr,
+  end
+                                    ... = a ^ d i * (∏ j, x₀ j ^ d j) : _
+                                    ... = (∏ j, x₀ j ^ d j) * a ^ d i : mul_comm _ _,
+  show (e a : σ → K) i ^ d i * (∏ (j : {j // j ≠ i}), (e a : σ → K) j ^ d j) =
+    a ^ d i * ∏ j, x₀ j ^ d j,
+  congr' 1,
+  { simp [e, equiv.subtype_preimage_symm_apply_coe_neg _ x₀ _ _ (not_not.mpr rfl)], },
+  { apply fintype.prod_congr,
+    rintros ⟨j, hj⟩,
+    show (e a : σ → K) j ^ d j = x₀ ⟨j, hj⟩ ^ d j,
+    simp [e, equiv.subtype_preimage_symm_apply_coe_pos _ x₀ _ _ hj], }
+end
+
+variables [decidable_eq K] [decidable_eq σ]
+
+/-- The Chevalley–Warning theorem.
+Let `(f i)` be a finite family of multivariate polynomials
+in finitely many variables (`X s`, `s : σ`) over a finite field of characteristic `p`.
+Assume that the sum of the total degrees of the `f i` is less than the cardinality of `σ`.
+Then the number of common solutions of the `f i` is divisible by `p`. -/
+theorem char_dvd_card_solutions_family (p : ℕ) [char_p K p]
+  {ι : Type*} {s : finset ι} {f : ι → mv_polynomial σ K}
+  (h : (∑ i in s, (f i).total_degree) < fintype.card σ) :
+  p ∣ fintype.card {x : σ → K // ∀ i ∈ s, (f i).eval x = 0} :=
+begin
+  have hq : 0 < q - 1, { rw [← card_units, fintype.card_pos_iff], exact ⟨1⟩ },
+  let S : finset (σ → K) := { x ∈ univ | ∀ i ∈ s, (f i).eval x = 0 },
+  have hS : ∀ (x : σ → K), x ∈ S ↔ ∀ (i : ι), i ∈ s → eval x (f i) = 0,
+  { intros x, simp only [S, true_and, sep_def, mem_filter, mem_univ], },
+  /- The polynomial `F = ∏ i in s, (1 - (f i)^(q - 1))` has the nice property
+  that it takes the value `1` on elements of `{x : σ → K // ∀ i ∈ s, (f i).eval x = 0}`
+  while it is `0` outside that locus.
+  Hence the sum of its values is equal to the cardinality of
+  `{x : σ → K // ∀ i ∈ s, (f i).eval x = 0}` modulo `p`. -/
+  let F : mv_polynomial σ K := ∏ i in s, (1 - (f i)^(q - 1)),
+  have hF : ∀ x, F.eval x = if x ∈ S then 1 else 0,
+  { intro x,
+    calc F.eval x = ∏ i in s, (1 - f i ^ (q - 1)).eval x :
+      by { convert eval₂_prod id _ _ _, exact is_semiring_hom.id }
```

Comment:
This should be (the missing) lemma `eval_prod`.

Reply:
Thanks, done